pybind11/how-to.rst

```diff
@@ -353,28 +348,47 @@ Wrapping *static* member functions is trivial:
 
 .. code-block:: cpp
 
-    clsExampleOne.def_static("initializeSomething", &ExampleOne::initializeSomething);
+    cls.def_static("initializeSomething", &ExampleOne::initializeSomething);
 
 Wrapping operators
 ^^^^^^^^^^^^^^^^^^
 
-According to our :ref:`rule on operators <style-guide-pybind11-operator>` we can either wrap
-operators directly, or use a lambda.
+According to our :ref:`rule on operators <style-guide-pybind11-operator>` we can either wrap operators directly, or use a lambda.
 Here we use both approaches:
 
 .. code-block:: cpp
 
-    clsExampleOne.def("__eq__", &ExampleOne::operator==, py::is_operator());
-    clsExampleOne.def("__ne__", &ExampleOne::operator!=, py::is_operator());
-    clsExampleOne.def("__iadd__", &ExampleOne::operator+= /* no py::is_operator() here */);
-    clsExampleOne.def("__add__", [](ExampleOne const & self, ExampleOne const & other) { return self + other; }, py::is_operator());
+    cls.def("__eq__", &ExampleOne::operator==, py::is_operator());
+    cls.def("__ne__", &ExampleOne::operator!=, py::is_operator());
+    cls.def("__iadd__", &ExampleOne::operator+= /* no py::is_operator() here */);
+    cls.def("__add__", [](ExampleOne const & self, ExampleOne const & other) { return self + other; }, py::is_operator());
 
 .. note::
 
     * We use ``py::is_operator()`` to return ``NotImplemented`` on failure.
     * We don't use ``py::is_operator()`` for in-place operators as this can lead to confusing behavior.
     * We name the lambda arguments :ref:`self <style-guide-pybind11-lambda-self-argument>` and :ref:`other <style-guide-pybind11-lambda-other-argument>`.
 
+Module-Level Declaration
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Module-level free functions and variables can be declared inside a ``wrapType`` callback, and should when they're closely related to the class it defines.
+
+In other cases, you can add a callback not associated with any class by calling the ``wrapFunctions`` method instead:
+
+.. code-block:: cpp
+
+    wrappers.wrapFunctions(
+        [](auto & mod) {
+            // any number of module-level wrappers go here
+        }
+    );
+
+.. note::
+
+    We do not attempt to update the ``__module__`` attribute of free functions, as these are rarely used.
+    Free functions for which ``__module__`` is important (e.g. functions passed to ``__reduce__`` for pickling) should have their ``__module__``modified manually.
```

Comment:
Space before "modified".  I'm not sure an "e.g." is sufficient for someone to know when this is important.

Reply:
I don't really have any good criteria, but I can say that I don't know how to reset `__module__` for all function wrappers automatically without adding a ton of boilerplate.  To be honest, it's probably just something that you don't realize you need until it bites you, and I'm not really sure what to do about that.

Reply:
Yes, both still work, but the former is can make pickle files that rely on the function unnecessarily fragile to later refactoring.  I'll rewrite to clearly explain the pickle example, as it's the only one I've actually got.